src/bin/cql-stress-cassandra-stress/operation/row_generator.rs

```diff
@@ -0,0 +1,147 @@
+use scylla::_macro_internal::CqlValue;
+
+use crate::{
+    java_generate::{
+        distribution::{fixed::FixedDistribution, sequence::SeqDistribution, Distribution},
+        values::{Blob, Generator, GeneratorConfig, HexBlob},
+    },
+    settings::CassandraStressSettings,
+};
+use std::sync::Arc;
+
+use super::recompute_seed;
+
+/// A row generator structure.
+///
+/// Row-generation logic:
+/// - sample the `pk_seed` from `pk_seed_distribution`
+/// - seed the `pk_generator` with sampled `pk_seed`
+/// - generate the partition key with `pk_generator`
+/// - compute the seed for the `column_generators` based on generated pk
+/// - generate the rest of the row (seeding the `column_generators` with computed seed)
+///
+/// I think it's a great place to address how read and write workloads cooperate.
+/// For reference, see: https://github.com/scylladb/cql-stress/pull/43#discussion_r1304274035.
+///
+/// At first, we need to notice that as long as `pk_seed_distribution` is a DETERMINISTIC distribution,
+/// the row-generation logic is also deterministic.
+/// The `pk_seed_distribution` is the one provided via CLI with either `-pop seq` or `-pop dist` option.
+/// Notice that `-pop seq=1..5` is short for `-pop dist=SEQ(1..5)`.
+///
+/// For example, each time we execute the command:
+/// ```
+/// ./cassandra-stress write n=100 -pop dist=SEQ(1..100)
+/// ```
+/// it will result in generating the exact same set of 100 rows (no matter the order of sampling - this distribution is shared across multiple threads).
+///
+/// Now, one can validate the inserted data with a read routine.
+/// Each of these commands will be successful (meaning, the data will be successfully validated):
+/// ```
+/// ./cassandra-stress read n=100 -pop dist=SEQ(1..100)
+/// ./cassandra-stress read n=100 -pop dist=UNIFORM(1..100)
+/// ./cassandra-stress read n=100 -pop dist=GAUSSIAN(30..70)
+/// ```
+///
+/// To be more precise: any read workload that samples the partition_key seeds from the distribution
+/// sampling the values from the subset of range 1..100 (one used in the write routine) will successfully validate the data.
+///
+/// This also means, that if we used a NON-DETERMINISTIC distribution e.g. `UniformDistribution` (which is by default seeded with current time as millis)
+/// in the write workload, most of the times, the read workload will result in a validation error.
+/// For example:
```

Comment:
Side note: it's technically possible to achieve determinism with non-seq distributions, and AFAIR we already do this in the s-b command - you can use the operation number as the seed to the random generator, and then use it to generate the partition key. I'm not sure why the original c-s doesn't do it...

Perhaps it's worth considering this as a future improvement.

Reply:
Good point, and to be quite honest, I'd really love to follow the approach you mentioned. I'm not sure what c-s wanted to achieve by being non-deterministic in the first place.

I'll save this conversation for the future (in the comment), and hopefully we will discuss it with @avelanarius .